app/frontend/src/routes.ts

```diff
@@ -15,17 +15,39 @@ export const confirmChangeEmailRoute = "/confirm-email";
 
 export const signupRoute = "/signup";
 
-export const userRoute = "/user";
-export const aboutRoute = `${userRoute}/about`;
-export const homeRoute = `${userRoute}/home`;
-export const referencesRoute = `${userRoute}/references`;
-export const favoritesRoute = `${userRoute}/favorites`;
-export const photosRoute = `${userRoute}/photos`;
-
-export const editProfileRoute = `${userRoute}/edit`;
-export const editHostingPreferenceRoute = `${userRoute}/edit-hosting`;
-
-export const routeToUser = (username: string) => `${userRoute}/${username}`;
+// user
+
+const userBaseRoute = "/user";
+export type UserTab = "about" | "home" | "references" | "favorites" | "photos";
+export type EditUserTab = Extract<UserTab, "about" | "home">;
+
+export const userRoute = `${userBaseRoute}/:userId?/:tab?`;
+export const editUserRoute = `${userBaseRoute}/edit/:tab?`;
+
+type UserRouteOptions<T extends UserTab> = {
+  username?: string;
+  tab?: T;
+  edit?: T extends EditUserTab ? true : T extends undefined ? true : never;
```

Comment:
Why is `edit` typed as `true` if it's undefined? Doesn't that mean they are not on the edit page and so edit can't be true?

Reply:
If T is undefined it means caller didn't pass any value (or undefined) for the 'tab' property. So the result should be a url for the default tab for profile page which is 'about'. 

About page does have an 'edit' mode so you are allowed to pass 'true' for the 'edit' prop.

However, you can also omit the 'edit' prop for the 'about' tab. Which will return the view. 

Passing 'false' for edit prop explicitly isn't allowed by any overload as you can omit the edit prop in that case.

Reply:
```suggestion
  edit?: T extends EditUserTab | undefined ? true : never;
```
I think this would also work? Makes this marginally easier to read :)

Reply:
Haha, definitly true! I'll change it of course :)